verilog/analysis/verilog_linter.cc

```diff
@@ -77,57 +89,249 @@ using verible::LintWaiver;
-    return 2;
+ protected:
+  virtual void HandleViolation(const verible::LintViolation& violation,
+                               absl::string_view base, absl::string_view path,
+                               absl::string_view url,
+                               absl::string_view rule_name) const override {
+    formatter_.FormatViolation(stream_, violation, base, path, url, rule_name);
+    // TODO: handle fix when HandleViolation is called directly
+    if (violation.autofixes.size() > 0 && fix_ != nullptr) {
+      if (fix_->AddEdits(violation.autofixes[0].Edits())) {
+        (*stream_) << " (autofixed)";
+      } else {
+        VLOG(0) << "The fix conflicts with previously applied fixes, ignoring.";
+      }
+    }
+    (*stream_) << std::endl;
   }
-  *stream << lint_stream.str();
-  if (!lint_stream.str().empty() && lint_fatal) {
-    return 1;
+
+  void PrintFix(std::ostream& stream, absl::string_view text,
+                const verible::AutoFix& fix) const {
+    std::string after = fix.Apply(text);
+    verible::LineDiffs diff(text, after);
+
+    verible::LineDiffsToUnifiedDiff(stream, diff, 1);
+  }
+  const LinterConfiguration* config_;
+  mutable verible::AutoFix* fix_;
+};
+
+class InteractiveViolationFixer : public ViolationFixer {
+ public:
+  using ViolationFixer::ViolationFixer;
+
+  virtual void HandleLintRuleStatuses(
+      const std::vector<LintRuleStatus>& statuses, absl::string_view base,
+      absl::string_view path) const override {
+    rule_answers_.clear();
+    answer_ = kMissingAnswer;
+    ViolationFixer::HandleLintRuleStatuses(statuses, base, path);
   }
-  return 0;
-}
+
+ protected:
+  virtual void HandleViolation(const verible::LintViolation& violation,
+                               absl::string_view base, absl::string_view path,
+                               absl::string_view url,
+                               absl::string_view rule_name) const override {
+    assert(fix_ != nullptr);
+
+    formatter_.FormatViolation(stream_, violation, base, path, url, rule_name);
+    (*stream_) << std::endl;
+
+    static absl::string_view help_message =
+        "y - apply fix\n"
+        "n - reject fix\n"
+        "a - apply this and all remaining fixes for violations of this rule\n"
+        "d - reject this and all remaining fixes for violations of this rule\n"
+        "A - apply this and all remaining fixes\n"
+        "D - reject this and all remaining fixes\n"
+        "p - show fix\n"
+        "P - show fixes applied so far\n"
+        "? - print this help and prompt again\n";
+
+    if (violation.autofixes.size() > 0) {
+      char c;
+      for (;;) {
+        if (answer_ != kMissingAnswer) {
+          c = answer_;
+        } else if (rule_answers_.find(rule_name) != rule_answers_.end()) {
+          c = rule_answers_.at(rule_name);
+        } else {
+          c = ReadChar("Autofix is available. Apply? [y,n,a,d,A,D,p,P,?] ");
+        }
```

Comment:
If we move the part that queries a user into a callback of sorts (roughly speaking `std::function<AnswerChoiceEnum(std::string_view formatted_violation)>`), then we can implement the user interaction as separate function, or also have an implementation that just always says yes...

Essentially what you already have here as `ViolationFixer` but as a callback function instead. It might using the linter in some programmatic way very simple as an implementation is just a closure away and less 'heavy' than implementing a class for it.

Having said that, I have not looked too closely at this file yet, this was just my firs thought how this might be implemented in a way that makes it pleasent to use programmatically. I might need to play a bit.

Reply:
Initial version done. I've also changed `VerilogLintTextStructure` return type. It now returns list of violation statuses without printing them, giving full control over violation handling if one needs it.